### PR TITLE
Feature/uncertainty updates

### PR DESCRIPTION
**Pull request by zebengberg**
### Features

- Add new `HistogramMatchingWithEckel` experimental humidity scaling model. This is still a work in progress.

### Breaking changes

- No longer attach empty fields "sdr", "rsr", "olr", "rf_sw", "rf_lw", "rf_net" onto the `source` parameter in `Cocip.eval` when the flight doesn't generate any persistent contrails.
- Remove params `humidity_scaling`, `rhi_adj_uncertainty`, and `rhi_boost_exponent_uncertainty` from `CocipUncertaintyParams`.

### Internals

- When possible, replace type hints `np.ndarray` -> `np.typing.NDArray[np.float_]` in the `cocip`, `cocip_params`, `cocip_uncertainty`, `radiative_forcing`, and `wake_vortex` modules.
- Slight performance enhancements in the `radiative_forcing` module.

## Tests

- [x] QC test passes locally (`make test`)
- [x] CI tests pass

## Reviewer

@mlshapiro

### CODE REVIEW

File: pycontrails/models/humidity_scaling.py [hunk 1]
```diff
@@ -553,10 +558,239 @@ def scale(
...
+    #: raised at model evaluation time. The order of the list must be
+    #: consistent with the order of the ERA5 ensemble members.
+    ensemble_specific_humidity: list[MetDataArray] | None = None
+
+    #: The specific member used. Must be in the range [0, 10). If None,
+    #: a ValueError will be raised at model evaluation time.
+    member: int | None = None
+
+
+class HistogramMatchingWithEckel(HumidityScaling):
```
Comment (mlshapiro) [inline pycontrails/models/humidity_scaling.py:670]: I would prefix this with `_` and put a note in the docstring that its experimental

```
.. warning ::

```

**Review by mlshapiro — APPROVED** [review]
Looks good - I want to talk about how we can use this after the uncertainty study, but this looks fine to me!

File: pycontrails/models/cocip/cocip.py [hunk 2]
```diff
@@ -376,8 +377,8 @@ def eval(
         # Save humidity scaling type to output attrs
         # NOTE: Do this after _process_flight because that method automatically
         # broadcasts all numeric source params.
-        if self.params["humidity_scaling"] is not None:
-            for k, v in self.params["humidity_scaling"].description.items():
+        if (hs := self.params["humidity_scaling"]) is not None:
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/models/cocip/cocip.py]: 95ac8d499acf3e77c9ea84d5850f395ae55758ac

File: pycontrails/models/humidity_scaling.py [hunk 3]
```diff
@@ -553,10 +558,239 @@ def scale(
...
         rhi_adj = np.interp(level, xp=xp, fp=fp)
 
         q = specific_humidity / rhi_adj
         rhi = thermo.rhi(q, air_temperature, air_pressure)
         return q, rhi
+
+
+@functools.cache
+def _load_iagos_quantiles() -> npt.NDArray[np.float64]:
+    path = pathlib.Path(__file__).parent / "quantiles" / "iagos_quantiles.npy"
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/models/humidity_scaling.py:572]: Yep, the `.npy` format is standard and stable.

I will also commit the files themselves (not yet included).

File: pycontrails/models/humidity_scaling.py [hunk 4]
```diff
@@ -553,10 +558,239 @@ def scale(
...
+    era5_quantiles = era5_quantiles[:, member]  # shape (801,)
+    iagos_quantiles = _load_iagos_quantiles()  # shape (801,)
+
+    return np.interp(era5_rhi, era5_quantiles, iagos_quantiles)
+
+
+def recalibrate_rhi(
+    era5_rhi_all_members: npt.NDArray[np.float_], member: int
+) -> npt.NDArray[np.float_]:
+    """Recalibrate ERA5-derived RHi values to IAGOS quantiles.
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/models/humidity_scaling.py:612]: Exactly! Eventually, once we compute the quantiles, we'll have a simpler `HistogramMatching` model which doesn't do any Eckel scaling and doesn't require all 10 ensembles.

File: pycontrails/models/humidity_scaling.py (hunk 1 above)
Comment (zebengberg) [reply pycontrails/models/humidity_scaling.py:678]: I included the warning, but not the prefix underscore!

4febb864eb18a0c6ec940d7535d2b92ca8f2812d

File: pycontrails/models/humidity_scaling.py [hunk 5]
```diff
@@ -553,10 +558,239 @@ def scale(
...
+
+    eckel_a = -0.005213832567192828
+    eckel_c = 2.7859172756970354
+
+    out = (ensemble_mean_rhi - eckel_a) + eckel_c * (recalibrated_rhi - ensemble_mean_rhi)
+    return out.astype(era5_rhi_all_members.dtype)
+
+
+@dataclasses.dataclass
+class HistogramMatchingWithEckelParams(ModelParams):
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/models/humidity_scaling.py:659]: 4febb864eb18a0c6ec940d7535d2b92ca8f2812d

File: pycontrails/models/cocip/cocip_uncertainty.py [hunk 6]
```diff
@@ -46,20 +47,25 @@ def rvs(self, *args: Any, **kwds: Any) -> np.ndarray:
...
+        if args or (kwds.get("size") is not None and kwds["size"] > 1):
             raise ValueError("habit_dirichlet distribution only supports creating one rv at a time")
 
-        default_distributions = CocipParams().habit_distributions
-        alpha_i = [0.5 + self.C * G_i for G_i in default_distributions]
-        distr = [stats.dirichlet(a) for a in alpha_i]
+        default_distributions = cocip_params._habit_distributions()
+        alpha_i = 0.5 + self.C * default_distributions
+
+        # In the first distribution, we assume all ice particles are spheres
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/models/cocip/cocip_uncertainty.py]: a304a2f883410ee549dab80b8062494e9a4cf3b9

File: pycontrails/models/humidity_scaling.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/models/humidity_scaling.py:572]: 5c2ba66c97e4ee06acc08ebd7ab8aa26c05079f7

File: pycontrails/models/cocip/cocip.py [hunk 7]
```diff
@@ -1274,15 +1278,6 @@ def _fill_empty_flight_results(self, return_list_flight: bool) -> Flight | list[
             Flight or list of Flight objects with empty variables.
         """
 
-        empty = np.full(shape=len(self.source), fill_value=np.nan)
-
-        self.source["sdr"] = empty.copy()
```
[the remark this answers is not in the document]
Comment (zebengberg) [reply pycontrails/models/cocip/cocip.py:1279]: You're right that we want consistency .... but these 6 variables are not attached to the source attribute even in the case when cocip predicts persistent contrails. Attaching these 6 empty variables is outdated. I think it's a bigger task to make the cocip outputs more consistent among all the different paths cocip could take and to better discuss the "verbose_outputs" parameter. I'll make an issue for this.

**Pull request merged**